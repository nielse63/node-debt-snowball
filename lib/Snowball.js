const Account = require('./Account');
const toCurrency = require('./helpers/toCurrency');

// steps of execution
// 1. sort accounts by interest rate and create Account objects
// 2. get sum of account balances
// 3. while sum of account balances is greater than 0...
// 1. for each account (pass in local additional payment var)...
// 1. calculate the accrued interest for the month
// 2. add the interest to the current account balance
// 3. add the min and additional payment amounts
// 4. if the total payment amount is greater than the current account balance, total payment is equal to account balance
// 5. subtract total payment amount from current account balance
// 6. return starting balance, ending balance, payment amount, interest accrued

class Snowball {
  accounts = [];
  additionalPayment = 0;
  balance = 0;
  currentAdditionalPayment = 0;
  paymentPlan = [];

  constructor(accounts, additionalPayment = 0) {
    this.setAccounts(accounts);
    const balance = this.getCurrentBalance();
    this.balance = balance;
    this.additionalPayment = additionalPayment;
    this.currentAdditionalPayment = additionalPayment;
  }

  setAccounts(accounts) {
    this.accounts = accounts
      .sort((a, b) => {
        if (a.interest > b.interest) return -1;
        if (a.interest < b.interest) return 1;
        return 0;
      })
      .map((account) => new Account(account));
    return this.accounts;
  }

  getCurrentBalance() {
    const sum = this.accounts
      .map(({ balance }) => balance)
      .reduce((a, b) => a + b, 0);
    return toCurrency(sum);
  }

  setNewAdditionalPayment(value) {
    const newAdditionalPayment =
      (this.currentAdditionalPayment || this.additionalPayment) - value;
    return newAdditionalPayment < 0 ? 0 : newAdditionalPayment;
  }

  makePaymentForAccount = (account) => {
    if (account.balance <= 0) {
      return {
        name: account.name,
        startingBalance: 0,
        endingBalance: 0,
        paymentAmount: 0,
        additionalPayment: 0,
      };
    }
    const additionalPayment = this.currentAdditionalPayment;
    const payment = account.makePayment(additionalPayment);

    this.currentAdditionalPayment = this.setNewAdditionalPayment(
      payment.additionalPayment
    );

    return {
      name: account.name,
      ...payment,
    };
  };

  makePaymentsForMonth() {
    this.currentAdditionalPayment = this.additionalPayment;
    const accounts = this.accounts.map(this.makePaymentForAccount);
    this.balance = this.getCurrentBalance();

    return {
      balance: this.balance,
      accounts,
    };
  }

  createPaymentPlan() {
    this.paymentPlan = [];

    while (this.balance > 0) {
      const payment = this.makePaymentsForMonth();
      this.paymentPlan.push(payment);
    }
    return this.paymentPlan;
  }
}

module.exports = Snowball;
