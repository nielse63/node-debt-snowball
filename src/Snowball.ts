import Account from './Account';
import Results from './Results';
import { toCurrency } from './helpers';
import type { AccountObject } from './types';

class Snowball {
  accounts: Account[] = [];
  additionalPayment: number;
  balanceStart: number;
  currentBalance: number;
  snowballAmount: number;

  constructor(accounts: AccountObject[], additionalPayment = 0) {
    this.accounts = this.setAccounts(accounts);
    this.balanceStart = this.getCurrentBalance();
    this.currentBalance = this.balanceStart;
    this.additionalPayment = additionalPayment;
    this.snowballAmount = additionalPayment;
  }

  parseAccounts(accounts: AccountObject[]) {
    if (!Array.isArray(accounts)) {
      throw new Error('accounts must be an array');
    }

    const defaultAccount = {
      name: '',
      interest: 0,
      balance: 0,
      minPayment: 0,
    };

    return accounts
      .filter((account) => {
        return account?.constructor === Object;
      })
      .map((account) => {
        return {
          ...defaultAccount,
          ...account,
        };
      });
  }

  setAccounts(accounts: AccountObject[]) {
    this.accounts = this.parseAccounts(accounts)
      .sort((a, b) => {
        if (a.interest > b.interest) return -1;
        if (a.interest < b.interest) return 1;
        return 0;
      })
      .map((account) => new Account(account));
    return this.accounts;
  }

  getCurrentBalance() {
    const sum = this.accounts
      .map(({ balance }) => balance)
      .reduce((a, b) => a + b, 0);
    return toCurrency(sum);
  }

  makePaymentForAccount = (account: Account) => {
    const payment = account.makePayment(this.snowballAmount);
    if (payment.balanceEnd > 0) {
      this.snowballAmount = 0;
    } else if (this.balanceStart > 0) {
      this.snowballAmount = toCurrency(
        this.additionalPayment + account.minPayment - payment.paymentAmount
      );
    }

    return {
      name: account.name,
      ...payment,
    };
  };

  makePaymentsForMonth() {
    this.snowballAmount = this.additionalPayment;
    const accounts = this.accounts.map(this.makePaymentForAccount);
    this.currentBalance = this.getCurrentBalance();

    return {
      balance: this.currentBalance,
      accounts,
    };
  }

  /**
   * Generates a payment plan
   * @returns {Payment[]} - An array of payments
   */
  createPaymentPlan() {
    const paymentPlan = [];

    while (this.currentBalance > 0) {
      const payment = this.makePaymentsForMonth();
      paymentPlan.push(payment);
    }
    return new Results(paymentPlan);
  }
}

export default Snowball;
