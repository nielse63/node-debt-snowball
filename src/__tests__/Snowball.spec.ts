import Snowball from '../Snowball';

const accounts = [
  {
    name: 'Credit Card',
    interest: 14.99,
    balance: 1000,
    minPayment: 75,
  },
  {
    name: 'Student Loan',
    interest: 4.75,
    balance: 7500,
    minPayment: 150,
  },
];
const additionalPayment = 100;

describe('Snowball', () => {
  let snowball;
  beforeEach(() => {
    snowball = new Snowball(accounts, additionalPayment);
  });

  it('should set default additionalPayment value', () => {
    snowball = new Snowball(accounts);
    expect(snowball.additionalPayment).toEqual(0);
  });

  describe('parseAccounts', () => {
    it('should throw error if accounts is not an array', () => {
      expect(() => snowball.parseAccounts('foo')).toThrow();
    });

    it('should filter non-object elements', () => {
      const output = snowball.parseAccounts(['foo', 'bar', 123, accounts[0]]);
      expect(output).toBeArrayOfSize(1);
      expect(output).toEqual([accounts[0]]);
    });

    it('should set default values in objects', () => {
      const output = snowball.parseAccounts([accounts[0], {}]);
      expect(output).toBeArrayOfSize(2);
      expect(output[1]).toEqual({
        name: '',
        interest: 0,
        balance: 0,
        minPayment: 0,
      });
    });

    it('should return an array of parsed account objects', () => {
      const output = snowball.parseAccounts(accounts);
      expect(output).toEqual([
        {
          name: 'Credit Card',
          interest: 14.99,
          balance: 1000,
          minPayment: 75,
        },
        {
          name: 'Student Loan',
          interest: 4.75,
          balance: 7500,
          minPayment: 150,
        },
      ]);
    });
  });

  describe('setAccounts', () => {
    it('should sort accounts by interest rate descending', () => {
      snowball = new Snowball([
        {
          name: 'Student Loan',
          interest: 6.55,
          balance: 1000,
          minPayment: 40,
        },
        {
          name: 'Credit Card',
          interest: 14.99,
          balance: 5000,
          minPayment: 120,
        },
      ]);
      expect(snowball.accounts[0].name).toEqual('Credit Card');
    });

    it('should maintain order if interest rates are the same', () => {
      snowball = new Snowball([
        {
          name: 'Student Loan',
          interest: 6.55,
          balance: 1000,
          minPayment: 40,
        },
        {
          name: 'Credit Card',
          interest: 6.55,
          balance: 5000,
          minPayment: 120,
        },
      ]);
      expect(snowball.accounts[0].name).toEqual('Student Loan');
    });
  });

  describe('getCurrentBalance', () => {
    it('should calculate sum balance correctly', () => {
      const balance = snowball.getCurrentBalance();
      expect(balance).toEqual(8500);
    });
  });

  describe('makePaymentForAccount', () => {
    let account;
    beforeEach(() => {
      [account] = snowball.accounts;
    });

    it('should return object with expected shape', () => {
      const payment = snowball.makePaymentForAccount(account);
      expect(payment).toBeObject();
      expect(payment.name).toEqual(account.name);
      expect(payment.startingBalance).toBeNumber();
      expect(payment.endingBalance).toBeNumber();
      expect(payment.paymentAmount).toBeNumber();
      expect(payment.additionalPayment).toBeNumber();
    });
  });

  describe('makePaymentsForMonth', () => {
    it('should return object with expected shape', () => {
      const results = snowball.makePaymentsForMonth();
      expect(results).toBeObject();
      expect(results.balance).toBeNumber();
      expect(results.accounts).toBeArray();
    });

    it('should calculate balance correctly', () => {
      const results = snowball.makePaymentsForMonth();
      const expected =
        results.accounts[0].endingBalance + results.accounts[1].endingBalance;
      expect(results.balance).toEqual(expected);
    });

    it('should update balance', () => {
      const oldBalance = snowball.startingBalance;
      const results = snowball.makePaymentsForMonth();
      expect(snowball.currentBalance).toBeLessThan(oldBalance);
      expect(results.balance).toEqual(snowball.currentBalance);
    });
  });

  describe('createPaymentPlan', () => {
    it('should return an array', () => {
      const paymentPlan = snowball.createPaymentPlan();
      expect(paymentPlan).toBeArray();
      expect(paymentPlan[0]).toBeObject();
    });

    it('should call makePaymentsForMonth', () => {
      const spy = jest.spyOn(snowball, 'makePaymentsForMonth');
      snowball.createPaymentPlan();
      expect(spy).toHaveBeenCalled();
    });

    it('should calculate correct balances', () => {
      const paymentPlan = snowball.createPaymentPlan();
      expect(paymentPlan[1].balance).toEqual(7931.85);
      expect(paymentPlan[4].accounts[0].paymentAmount).toEqual(175);
      expect(paymentPlan[5].accounts[0].paymentAmount).toEqual(168.99);
      expect(paymentPlan[6].accounts).toBeArrayOfSize(1);
    });

    it('should match snapshot', () => {
      const paymentPlan = snowball.createPaymentPlan();
      expect(paymentPlan).toMatchSnapshot();
    });
  });
});
