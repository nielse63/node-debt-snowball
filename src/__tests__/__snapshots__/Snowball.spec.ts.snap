// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snowball createPaymentPlan should match snapshot 1`] = `
[
  {
    "accounts": [
      {
        "accruedInterest": 12.49,
        "additionalPayment": 100,
        "endingBalance": 837.49,
        "name": "Credit Card",
        "paymentAmount": 175,
        "startingBalance": 1000,
      },
      {
        "accruedInterest": 29.69,
        "additionalPayment": 0,
        "endingBalance": 7379.69,
        "name": "Student Loan",
        "paymentAmount": 150,
        "startingBalance": 7500,
      },
    ],
    "balance": 8217.18,
  },
  {
    "accounts": [
      {
        "accruedInterest": 10.46,
        "additionalPayment": 100,
        "endingBalance": 672.95,
        "name": "Credit Card",
        "paymentAmount": 175,
        "startingBalance": 837.49,
      },
      {
        "accruedInterest": 29.21,
        "additionalPayment": 0,
        "endingBalance": 7258.9,
        "name": "Student Loan",
        "paymentAmount": 150,
        "startingBalance": 7379.69,
      },
    ],
    "balance": 7931.85,
  },
  {
    "accounts": [
      {
        "accruedInterest": 8.41,
        "additionalPayment": 100,
        "endingBalance": 506.36,
        "name": "Credit Card",
        "paymentAmount": 175,
        "startingBalance": 672.95,
      },
      {
        "accruedInterest": 28.73,
        "additionalPayment": 0,
        "endingBalance": 7137.63,
        "name": "Student Loan",
        "paymentAmount": 150,
        "startingBalance": 7258.9,
      },
    ],
    "balance": 7643.99,
  },
  {
    "accounts": [
      {
        "accruedInterest": 6.33,
        "additionalPayment": 100,
        "endingBalance": 337.69,
        "name": "Credit Card",
        "paymentAmount": 175,
        "startingBalance": 506.36,
      },
      {
        "accruedInterest": 28.25,
        "additionalPayment": 0,
        "endingBalance": 7015.88,
        "name": "Student Loan",
        "paymentAmount": 150,
        "startingBalance": 7137.63,
      },
    ],
    "balance": 7353.57,
  },
  {
    "accounts": [
      {
        "accruedInterest": 4.22,
        "additionalPayment": 100,
        "endingBalance": 166.91,
        "name": "Credit Card",
        "paymentAmount": 175,
        "startingBalance": 337.69,
      },
      {
        "accruedInterest": 27.77,
        "additionalPayment": 0,
        "endingBalance": 6893.65,
        "name": "Student Loan",
        "paymentAmount": 150,
        "startingBalance": 7015.88,
      },
    ],
    "balance": 7060.56,
  },
  {
    "accounts": [
      {
        "accruedInterest": 2.08,
        "additionalPayment": 100,
        "endingBalance": 0,
        "name": "Credit Card",
        "paymentAmount": 168.99,
        "startingBalance": 166.91,
      },
      {
        "accruedInterest": 27.29,
        "additionalPayment": -93.99,
        "endingBalance": 6864.93,
        "name": "Student Loan",
        "paymentAmount": 56.01,
        "startingBalance": 6893.65,
      },
    ],
    "balance": 6864.93,
  },
  {
    "accounts": [
      {
        "accruedInterest": 27.17,
        "additionalPayment": 175,
        "endingBalance": 6567.1,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 6864.93,
      },
    ],
    "balance": 6567.1,
  },
  {
    "accounts": [
      {
        "accruedInterest": 25.99,
        "additionalPayment": 175,
        "endingBalance": 6268.09,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 6567.1,
      },
    ],
    "balance": 6268.09,
  },
  {
    "accounts": [
      {
        "accruedInterest": 24.81,
        "additionalPayment": 175,
        "endingBalance": 5967.9,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 6268.09,
      },
    ],
    "balance": 5967.9,
  },
  {
    "accounts": [
      {
        "accruedInterest": 23.62,
        "additionalPayment": 175,
        "endingBalance": 5666.52,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 5967.9,
      },
    ],
    "balance": 5666.52,
  },
  {
    "accounts": [
      {
        "accruedInterest": 22.43,
        "additionalPayment": 175,
        "endingBalance": 5363.95,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 5666.52,
      },
    ],
    "balance": 5363.95,
  },
  {
    "accounts": [
      {
        "accruedInterest": 21.23,
        "additionalPayment": 175,
        "endingBalance": 5060.18,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 5363.95,
      },
    ],
    "balance": 5060.18,
  },
  {
    "accounts": [
      {
        "accruedInterest": 20.03,
        "additionalPayment": 175,
        "endingBalance": 4755.21,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 5060.18,
      },
    ],
    "balance": 4755.21,
  },
  {
    "accounts": [
      {
        "accruedInterest": 18.82,
        "additionalPayment": 175,
        "endingBalance": 4449.03,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 4755.21,
      },
    ],
    "balance": 4449.03,
  },
  {
    "accounts": [
      {
        "accruedInterest": 17.61,
        "additionalPayment": 175,
        "endingBalance": 4141.64,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 4449.03,
      },
    ],
    "balance": 4141.64,
  },
  {
    "accounts": [
      {
        "accruedInterest": 16.39,
        "additionalPayment": 175,
        "endingBalance": 3833.03,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 4141.64,
      },
    ],
    "balance": 3833.03,
  },
  {
    "accounts": [
      {
        "accruedInterest": 15.17,
        "additionalPayment": 175,
        "endingBalance": 3523.2,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 3833.03,
      },
    ],
    "balance": 3523.2,
  },
  {
    "accounts": [
      {
        "accruedInterest": 13.95,
        "additionalPayment": 175,
        "endingBalance": 3212.15,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 3523.2,
      },
    ],
    "balance": 3212.15,
  },
  {
    "accounts": [
      {
        "accruedInterest": 12.71,
        "additionalPayment": 175,
        "endingBalance": 2899.86,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 3212.15,
      },
    ],
    "balance": 2899.86,
  },
  {
    "accounts": [
      {
        "accruedInterest": 11.48,
        "additionalPayment": 175,
        "endingBalance": 2586.34,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 2899.86,
      },
    ],
    "balance": 2586.34,
  },
  {
    "accounts": [
      {
        "accruedInterest": 10.24,
        "additionalPayment": 175,
        "endingBalance": 2271.58,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 2586.34,
      },
    ],
    "balance": 2271.58,
  },
  {
    "accounts": [
      {
        "accruedInterest": 8.99,
        "additionalPayment": 175,
        "endingBalance": 1955.57,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 2271.58,
      },
    ],
    "balance": 1955.57,
  },
  {
    "accounts": [
      {
        "accruedInterest": 7.74,
        "additionalPayment": 175,
        "endingBalance": 1638.31,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 1955.57,
      },
    ],
    "balance": 1638.31,
  },
  {
    "accounts": [
      {
        "accruedInterest": 6.48,
        "additionalPayment": 175,
        "endingBalance": 1319.79,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 1638.31,
      },
    ],
    "balance": 1319.79,
  },
  {
    "accounts": [
      {
        "accruedInterest": 5.22,
        "additionalPayment": 175,
        "endingBalance": 1000.01,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 1319.79,
      },
    ],
    "balance": 1000.01,
  },
  {
    "accounts": [
      {
        "accruedInterest": 3.96,
        "additionalPayment": 175,
        "endingBalance": 678.97,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 1000.01,
      },
    ],
    "balance": 678.97,
  },
  {
    "accounts": [
      {
        "accruedInterest": 2.69,
        "additionalPayment": 175,
        "endingBalance": 356.66,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 678.97,
      },
    ],
    "balance": 356.66,
  },
  {
    "accounts": [
      {
        "accruedInterest": 1.41,
        "additionalPayment": 175,
        "endingBalance": 33.07,
        "name": "Student Loan",
        "paymentAmount": 325,
        "startingBalance": 356.66,
      },
    ],
    "balance": 33.07,
  },
  {
    "accounts": [
      {
        "accruedInterest": 0.13,
        "additionalPayment": 175,
        "endingBalance": 0,
        "name": "Student Loan",
        "paymentAmount": 33.2,
        "startingBalance": 33.07,
      },
    ],
    "balance": 0,
  },
]
`;
